File: src/AD.BaseTypes.Core/BaseType.cs
using System.Diagnostics.CodeAnalysis;

namespace AD.BaseTypes;

/// <summary>
/// Base type.
/// </summary>
/// <typeparam name="TBaseType">The base type.</typeparam>
/// <typeparam name="TWrapped">The wrapped type.</typeparam>
public static class BaseType<TBaseType, TWrapped> where TBaseType : IBaseType<TBaseType, TWrapped> where TWrapped : notnull
{
    /// <summary>
    /// Creates the base type.
    /// </summary>
    /// <param name="value">The wrapped value.</param>
    /// <returns>The created base type.</returns>
    /// <exception cref="ArgumentException">The parameter <paramref name="value"/> is invalid.</exception>
    public static TBaseType From(TWrapped value) => TBaseType.From(value);

    /// <summary>
    /// Tries to create a base type.
    /// </summary>
    /// <param name="value">The wrapped value.</param>
    /// <param name="baseType">The created base type.</param>
    /// <param name="errorMessage">The error message.</param>
    /// <returns>True, if the base type is created.</returns>
    public static bool TryFrom(TWrapped value, [MaybeNullWhen(false)] out TBaseType baseType, [MaybeNullWhen(true)] out string errorMessage)
    {
        try
        {
            baseType = From(value);
            errorMessage = default;
            return true;
        }
        catch (ArgumentException ex)
        {
            baseType = default;
            errorMessage = ex.Message;
            return false;
        }
    }
}

File: src/AD.BaseTypes.FSharp/BaseType.fs
/// Module for base types.
module AD.BaseTypes.FSharp.BaseType

open AD.BaseTypes

/// <summary>
/// Creates a base type.
/// </summary>
/// <param name="value">The base type's value.</param>
/// <returns>The created base type or an error message.</returns>
let create<'baseType, 'wrapped when 'baseType :> IBaseType<'baseType, 'wrapped>> value =
    match value |> BaseType<'baseType, _>.TryCreate with
    | true, baseType, _ -> baseType |> Ok
    | false, _, error -> error |> Error

/// Gets the base type's value.
let value (baseType : #IBaseType<_>) = baseType.Value

/// Gets the base type's value.
let (|Value|) (baseType : #IBaseType<_>) = baseType.Value

/// Binds a base type.
let bind binder (baseType : 'baseType) =
    baseType |> value |> binder |> Result.bind create<'baseType, _>

/// Maps a base type.
let map mapper (baseType : 'baseType) =
    baseType |> value |> mapper |> create<'baseType, _>

/// Maps the base type's value.
let mapValue mapper = value >> mapper
